docs/request_handling.md

```diff
@@ -0,0 +1,54 @@
+# Request handling.
+
+To handle requests sent by client, the nodes require a ledger and/or a state as a data store. Request handling logic is written in `RequestHandler`. 
+`RequestHandler` is extended by new classes to support new requests. A node provides methods `register_new_ledger` to register new `Ledger` and `register_req_handler` to register new `RequestHandler`s.
+There can be a callback registered to execute after a write request has been successfully executed using `register_executer`.
+
+  
+There are 2 types of requests a client can send:
+-   Query:
+    Here the client is requesting transactions or some state variables from the node(s). The client can either send a `GET_TXN` to get any transaction with a sequence number from any ledger.
+    Or it can send specific `GET_*` transactions for queries
+-   Write:
+    Here the client is asking the nodes to write a new transaction to the ledger and change some state variable. This requires the nodes to run a consensus protocol (currently RBFT).
+    If the protocol run is successful, then the client's proposed changes are written. 
+
+
+#### Request handling
+Below is a description of how a request is processed.
+A node on receiving a client request in  `validateClientMsg`:
+-   The node performs static validation checks (validation that does not require any state, like mandatory fields are present, etc), it calls `RequestHandler`'s `doStaticValidation`
+-   If the static validation passes, it checks if the signature check is required and does that if needed in `verifySignature`. More on this later.
+-   Checks if it a generic transaction query (`GET_TXN`). If it is then query the ledger for that particular sequence number and return the result. A `REQNACK` might be sent if the query is not correctly constructed. 
+-   Checks if it a specific query (needs a specific `RequestHandler`), if it is then it uses `process_query` that uses the specific `RequestHandler` to respond to the query. A `REQNACK` might be sent if the query is not correctly constructed.
+-   If it is a write, then node checks if it has already processed the request before by checking the uniqueness of `identifier` and `reqId` fields of the Request.
+    -   If it has already processed the request, then it sends the corresponding `Reply` to the client
+    -   If the `Request` is already in process, then it sends an acknowledgement to the client as a `REQACK`
+    -   If the node has not seen the `Request` before it broadcasts the `Request` to all nodes in a `PROPAGATE`.
+    -   Once a node receives sufficient (`Quorums.propagate`) `PROPAGATE`s for a request, it forwards the request to its replicas.
+    -   A primary replica on receiving a forwarded request does dynamic validation (requiring state, like if violating some unique constraint or doing an unauthorised action) 
```

Comment:
Should we mention batching here? That Primary applies dynamic validation for each request in the batch, and that the batch can still be processed if there are invalid requests there. Reject is sent just for invalid requests from the batch,

Reply:
Batch is an optimisation for consensus, i didn't want to digress, i think a detailed doc explaining our differences in 3PC with RBFT should list batching, merkle roots and checks in 3PC messages, etc

Reply:
Ok, that's fine. Probably it's sufficient here to just mention this doc so that one can have a look how this all works with batches.